spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -1017,4 +1019,41 @@ public String unknown(
       return String.format("%s(%s) %s %s", transform, sourceName, direction, nullOrder);
     }
   }
+
+  public static Map<String, String> buildWriteOptions(String branch, LogicalWriteInfo info) {
+    ImmutableMap.Builder<String, String> builder =
+        ImmutableMap.<String, String>builder().putAll(info.options());
+
+    if (branch != null) {
+      String optionBranch = info.options().get(SparkReadOptions.BRANCH);
+      if (optionBranch == null) {
+        builder.put(SparkWriteOptions.BRANCH, branch);
+      } else if (!optionBranch.equals(branch)) {
+        throw new ValidationException(
+            "Cannot override branch to write more than once, received %s in identifier and %s in write options",
+            branch, optionBranch);
+      }
+    }
+
+    return builder.build();
+  }
+
+  public static CaseInsensitiveStringMap buildScanOptions(
+      String branch, CaseInsensitiveStringMap options) {
+    ImmutableMap.Builder<String, String> builder =
+        ImmutableMap.<String, String>builder().putAll(options);
+
+    if (branch != null) {
+      String optionBranch = options.get(SparkReadOptions.BRANCH);
+      if (optionBranch == null) {
+        builder.put(SparkReadOptions.BRANCH, branch);
+      } else if (!optionBranch.equals(branch)) {
+        throw new ValidationException(
+            "Cannot override branch to read more than once, received %s in identifier and %s in scan options",
```

Comment:
Yeah so I think this is a point of discussion, my take was we deprecate the SparkReadOptions just so from a user experience perspective there's not a bunch of ways to query the head of a branch. Every operation (SQL reads and writes or Dataframe) has the branch in the identifier. The only addition for SQL is the time travel syntax.

Reply:
I agree that we should still keep spark read options. Waiting for others to reply.

Reply:
But we also have other names like `.at_timestamp_`, `.snapshot_id_`, `tag_`, which maps to the read options. So I think it's hard to argue that we don't do that for branch.

Reply:
I think it would be cleaner to just pass the branch directly into `SparkReadConf` and detect when to use it, rather than creating a copy of `CaseInsensitiveStringMap`.

Since we've already shipped with `branch` and `tag` options, we should leave them in. Spark considers those part of the table definition, so I think that it is safe to continue to allow them. But because of the issues with needing to produce scans based on write options, we should avoid adding places that depend on them.